### PR TITLE
Handle multi anonymous struct with same USR

### LINKED ISSUE #558: Generated dart code for unions/structs containing anonymous structs is broken
## Steps to reproduce

Here's a C header file that produces a bad ffigen binding:

```c
typedef struct s1 { int i; } s1;
typedef struct s2 { int i; } s2;

typedef union fail1 {
  struct {
      struct s1 v1;
  };
  struct {
      struct s2 v2;
  };
} fail1;
```

Generate the dart binding:

```
╰─➤  flutter pub run ffigen --config ffigen.yaml                                         
Running in Directory: '...'
Input Headers: [./src/ffigenbug.h]
Finished, Bindings generated in .../ffigenbug/./lib/ffigenbug_bindings_generated.dart
```

<details>
  <summary>
The ffigen.yml is the basic one from the flutter plugin template, nothing important:
</summary>

```yaml
# Run with `flutter pub run ffigen --config ffigen.yaml`.
name: FfigenbugBindings
description: |
  Bindings for `src/ffigenbug.h`.

  Regenerate bindings with `flutter pub run ffigen --config ffigen.yaml`.
output: 'lib/ffigenbug_bindings_generated.dart'
headers:
  entry-points:
    - 'src/ffigenbug.h'
  include-directives:
    - 'src/ffigenbug.h'
preamble: |
  // ignore_for_file: always_specify_types
  // ignore_for_file: camel_case_types
  // ignore_for_file: non_constant_identifier_names
comments:
  style: any
  length: full
```
</details>

## Obtained result

The generated dart file is:

```dart
import 'dart:ffi' as ffi;

class s1 extends ffi.Struct {
  @ffi.Int()
  external int i;
}

class s2 extends ffi.Struct {
  @ffi.Int()
  external int i;
}

class fail1 extends ffi.Union {
  external UnnamedStruct1 unnamed;

  external UnnamedStruct1 unnamed1;
}

class UnnamedStruct1 extends ffi.Struct {
  external s1 v1;
}
```

## Expected result

First of all, this is wrong because the union contains the same struct twice instead of two different structs.

Additionally, and this is more of a feature request, it would be nice if ffigen could flatten the anonymous structs if they contain only one member like:

```dart
class fail1 extends ffi.Union {
  external s1 v1;

  external s2 v2;
}
```

## Additional information

ffigen 7.2.10
The C header file in our real case is generated from rust code by cbindgen, we don't have much control over its structure.
I just noticed that using a struct instead of a union exposes the same bug.

### PR DESCRIPTION
**Pull request by mannprerak2**
Closes #558 

- Add the source file offset suffix to anonymous declaration's USR.
- Add test, update changelog

@dcharkes you may wanna move the stable branch up till the latest stable ffigen version `7.2.10` first. (I've added by commits on top of that commit - fe870e8f012361cf82a6f74e7b6adbba6f2a24fa)

### CODE REVIEW

**Comment by mannprerak2** [comment]
> I'll review the diff once it's rebased
Done. @dcharkes

File: lib/src/header_parser/utils.dart [hunk 1]
```diff
@@ -124,6 +128,15 @@ extension CXCursorExt on clang_types.CXCursor {
     return s;
   }
 
+  int sourceFileOffset() {
+    final cxsource = clang.clang_getCursorLocation(this);
+    final cxOffset = calloc<UnsignedInt>();
```
Comment (dcharkes) [inline lib/src/header_parser/utils.dart:133]: This isn't freed?

**Review by dcharkes — APPROVED** [review]
lgtm with comment

Thanks @mannprerak2 ! 🙏

File: lib/src/header_parser/utils.dart (hunk 1 above)
Comment (mannprerak2) [reply lib/src/header_parser/utils.dart:133]: Ugh, rookie mistake. 

Fixed it.

**Pull request merged**